snap: snap help output refresh

degville opened the pull request:
Proposed update to `snap help` output now that the Other category is too long. I touched a couple of other things too that I thought needed some attention:
- replaced the _Commands_ category with _Paths_ as this seems clearer to me 
  (Commands jarred with our use of command in help output)
- replaced _Other_ with _Status_ to hold commands for managing information/acknowledgements
- created a _Devices_ category to hold commands that can operate on the entire device
- some shuffling of commands from the old Other category (create-cohort, ack, recovery, reboot)
- put commands for each category into alphabetical order. There is some advantage in having common commands first, but our list had become seemingly random and difficult to scan.

Original `snap help` output:
```
         Basics: find, info, install, list, remove
        ...more: refresh, revert, switch, disable, enable
        History: changes, tasks, abort, watch
        Daemons: services, start, stop, restart, logs
       Commands: alias, aliases, unalias, prefer
  Configuration: get, set, unset, wait
        Account: login, logout, whoami
    Permissions: connections, interface, connect, disconnect
      Snapshots: saved, save, check-snapshot, restore, forget
          Other: version, warnings, okay, ack, known, model, create-cohort, recovery, reboot
    Development: run, pack, try, download, prepare-image
```

Proposed `snap help` output:
```
         Basics: find, info, install, list, refresh, remove
        ...more: create-cohort, disable, enable, revert, switch
        History: abort, changes, tasks, watch
        Daemons: logs, restart, services, start, stop
          Paths: alias, aliases, prefer, unalias
  Configuration: get, set, unset, wait
        Account: login, logout, whoami
    Permissions: connect, connections, disconnect, interface
      Snapshots: check-snapshot, forget, restore, save, saved
         Status: known, okay, version, warnings
        Devices: model, prepare-image, reboot, recovery
    Development: ack, download, pack, run, try
```

anonymouse64 commented:
A couple of small comments, but overall I like this a lot more, thanks for this!

> ```
>           Paths: alias, aliases, prefer, unalias
> ```

I think this would be more clear if it was _Apps_ because these snap commands are manipulating app aliases for snaps, the commands aren't really doing anything with paths from a high level (though I suppose from an implementation standpoint we are adjusting /snap/bin, which is a path but that's a bit of a weird angle to name the category I think)

> ```
>           Devices: model, prepare-image, reboot, recovery
> ```

While `prepare-image` is used for creating device images, it is never really meant to be called directly by users (and not actually on an Ubuntu Core device itself) and instead is a "routine" kind of command we expose primarily to be used/driven by ubuntu-image, so IMHO we should actually just drop it from the documented `snap` commands

degville commented:
> >           Paths: alias, aliases, prefer, unalias
> 
> I think this would be more clear if it was _Apps_ because these snap commands are manipulating app aliases for snaps, the commands aren't really doing anything with paths from a high level (though I suppose from an implementation standpoint we are adjusting /snap/bin, which is a path but that's a bit of a weird angle to name the category I think)

:+1: I think you're right about Paths, and _Apps_ is definitely better than _Commands_, and I really can't think of anything better for this that quickly communicates what the alias commands do - Apps and aliases on the same line is probably enough to imply their function if you know what you're looking for..

> >           Devices: model, prepare-image, reboot, recovery
> While `prepare-image` is used for creating device images, it is never really meant to be called directly by users (and not actually on an Ubuntu Core device itself) and instead is a "routine" kind of command we expose primarily to be used/driven by ubuntu-image, so IMHO we should actually just drop it from the documented `snap` commands

:+1: dropping prepare-image sounds like a good plan to me.

anonymouse64 commented:
> +1 I think you're right about Paths, and Apps is definitely better than Commands, and I really can't think of anything better for this that quickly communicates what the alias commands do - Apps and aliases on the same line is probably enough to imply their function if you know what you're looking for..

Actually now that you mention it maybe "Aliases" is a better name for the category since that's actually all these commands are about

degville commented:
> Actually now that you mention it maybe "Aliases" is a better name for the category since that's actually all these commands are about

I initially discounted _Aliases_ because I thought it looked weird and didn't provide much context:

```
        Daemons: services, start, stop, restart, logs
        Aliases: alias, aliases, unalias, prefer
  Configuration: get, set, unset, wait
```

...but I think I agree with you; it does what it says and doesn't look too bad.

anonymouse64 commented:
> I initially discounted Aliases because I thought it looked weird and didn't provide much context:

It is a little duplicative, but perhaps clarity is more important than duplicity here

bboozzoo commented:
I'm not quite sure about `create-cohort` in `more` section. The way I understand it, more is an extension of the basic set of commands, made to have the output fit within a reasonable number of columns. Maybe we could have a section like this at the end?
```
        Advanced: create-cohort
```

mvo5 commented on cmd/snap/cmd_help.go:
```diff
@@ -208,19 +208,23 @@ var helpCategories = []helpCategory{
+	}, {
+		Label:       i18n.G("Devices"),
+		Description: i18n.G("system management"),
+		Commands:    []string{"model", "prepare-image", "reboot", "recovery"},
```
I agree with Ian here, maybe we can hide "prepare-image", it's really a bit internal.

mvo5 reviewed (commented):
Thanks, looking super close. I would like to merge this soon, the thing I would consider changing it "Paths" vs "Commands" and then we can always iterate. But some more suggestions inline that I feel are less important.

zyga commented:
> > I initially discounted Aliases because I thought it looked weird and didn't provide much context:
> 
> It is a little duplicative, but perhaps clarity is more important than duplicity here

I would call it App Aliases, otherwise it's unclear what is being aliased.

zyga commented on cmd/snap/cmd_help.go:
```diff
@@ -180,23 +180,23 @@ var helpCategories = []helpCategory{
 		Label:       i18n.G("...more"),
 		Description: i18n.G("slightly more advanced snap management"),
-		Commands:    []string{"refresh", "revert", "switch", "disable", "enable"},
+		Commands:    []string{"create-cohort", "disable", "enable", "revert", "switch"},
```
I would put revert immediately after refresh and install paired with remove. I find this more natural than the somewhat arbitrary alphabetic order. I also agree with mvo, create-cohort should be the last element of this line.

zyga commented on cmd/snap/cmd_help.go:
```diff
@@ -180,23 +180,23 @@ var helpCategories = []helpCategory{
+		Commands:    []string{"logs", "restart", "services", "start", "stop"},
 	}, {
-		Label:       i18n.G("Commands"),
+		Label:       i18n.G("Paths"),
```
As mentioned earlier, I think this should be `App Aliases`

zyga reviewed (approved):
+1 on making the changes, suggested something myself as this is a bike shed dream topic :)

pedronis commented:
Thanks, some comments

* renaming Commands to Paths is confusing, I think it should stay Commands (or Aliases)
* given we are running out of lines, I don't think Status is useful (also other things are status), we should probably put back
   Other but keep the new Devices category. Splitting "ack" and "known" is odd as well,  also "ack" is not  a Development-only command
* sorting lexicographically is more consistent but is doing a disservice to clarity I think,  for anything but Other we should probably
  go back to importance, sequence of use sorting
* We should probably change the code to allow us to have more commands shown by snap help --all vs snap help, One solution could be to  have  "Special purpose" commands category that shows up only in snap help --all  and we would put prepare-image and create-cohort there

with hindsight the order of the lines for the categories themselves is a bit odd, it's also not alphabetical and could use some tweaking,  Permissions and Configuration might come before Commands perhaps

degville commented:
Thanks for all the feedback!

> * renaming Commands to Paths is confusing, I think it should stay Commands (or Aliases)

I feel quite strongly that Commands is wrong in the current context. My problem is that I don't think it's at all clear what we mean by _Commands_. In the truncated help output, we say the following:

```
The snap command lets you install, configure, refresh and remove snaps.
Snaps are packages that work across many different Linux distributions,
enabling secure delivery and operation of the latest apps and utilities.

Usage: snap <command> [<options>...]

Commands can be classified as follows:
```
That's three different uses of the word command and I think the user asking for help is going to find the Commands category confusing after the intro. I'm happy with Aliases, or App Aliases as Zyga suggests (or anything, really, this is just a suggestion).

> * given we are running out of lines, I don't think Status is useful (also other things are status), we should probably put back

I'm really just trying to avoid _Other_, which is maybe out of the scope of this PR and we can postpone for further comment. My problem with Other is that I think it looks lazy (imo, we can't be bothered to think of a category, so here you go) and it doesn't actually help the user - `snap help --all | grep thing` is more useful, so why put it in Other?

>   go back to importance, sequence of use sorting

Yep, ok. Maybe we should mention this in the help output to give users a clue as to why one command is ahead of another.

> with hindsight the order of the lines for the categories themselves is a bit odd, it's also not alphabetical and could use some tweaking, Permissions and Configuration might come before Commands perhaps

Yeah, I agree. 

> * We should probably change the code to allow us to have more commands shown by snap help --all vs snap help, One solution could be to  have  "Special purpose" commands category that shows up only in snap help --all  and we would put prepare-image and create-cohort there
> 

I think this is the best overall solution to these problems.

pedronis commented:
> given we are running out of lines, I don't think Status is useful (also other things are status), we should probably put back
I'm really just trying to avoid Other, which is maybe out of the scope of this PR and we can postpone for further comment. My problem with Other is that I think it looks lazy (imo, we can't be bothered to think of a category, so here you go) and it doesn't actually help the user - snap help --all | grep thing is more useful, so why put it in Other?

The main reason we need Other is that we cannot add more lines without breaking the it kind of fits into a default-sized terminal. It's not that we can't be bothered to find appropriate categories, we can, we would run out of lines though. And at least some of those commands are useful enough that appearing only in --all would be a problem

degville commented:
Thanks for the discussion and feedback! I've pushed the following changes:
- reinstated Other
- reinstated the most-use order for the commands in each category
- renamed Paths to App Aliases
- removed prepare-image

Proposed output is now as follows:

```
         Basics: find, info, install, remove, list
        ...more: refresh, revert, switch, disable, enable
        History: changes, tasks, abort, watch
        Daemons: services, start, stop, restart, logs
    Permissions: connections, interface, connect, disconnect
  Configuration: get, set, unset, wait
    App Aliases: alias, aliases, unalias, prefer
        Account: login, logout, whoami
      Snapshots: saved, save, check-snapshot, restore, forget
        Devices: model, reboot, recovery
          Other: version, warnings, okay, ack, known, create-cohort
    Development: download, pack, run, try
```

mvo5 reviewed (approved):
I like it

pedronis commented:
@degville I had to reinstate prepare-image for now because a unit test fails otherwise. I also tweaked the Device (was Devices) category a bit.

I'll do a follow-up to explore the things we discussed.

mvo5 reviewed (commented):
Still looks good to me
